src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    internal static partial class ClrThreadPool
+    {
+        private static WaitThreadNode s_waitThreadsHead = new WaitThreadNode
+        {
+            Thread = new WaitThread()
+        };
+
+        internal static void RegisterWaitHandle(RegisteredWaitHandle handle)
+        {
+            WaitThreadNode prev;
+            WaitThreadNode current = s_waitThreadsHead;
+            do
+            {
+                if(current.Thread.RegisterWaitHandle(handle))
+                {
+                    return;
+                }
+                prev = current;
+                current = current.Next;
+            } while (current != null);
+
+            prev.Next = new WaitThreadNode
+            {
+                Thread = new WaitThread()
+            };
+            prev.Next.Thread.RegisterWaitHandle(handle);
```

Comment:
I'm trying to decide now where to put the lock. I see two options:
1. Around the whole function body: This would be the simplest and the code would look clean; however, I do not know if the restriction of only one thread being able to register a wait at a time is too much of a perf hit.
2. Around the prev.Next assignment: I could put the lock here; however, this would either require me to again to a search further down the linked list (if another thread creates a prev.Next itself and another as well if this thread does not get scheduled) again to add the thread. Or I could create the wait thread before assigning and then search through the rest of the list to find a prev.Next to assign to that is null, but that might result in many more wait threads than needed.


I'm leaning towards 1, but if you have any suggestions please let me know.